api/src/main/java/io/cloudevents/visitor/CloudEventAttributesVisitor.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2018-Present The CloudEvents Authors
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.cloudevents.visitor;
+
+import io.cloudevents.types.Time;
+
+import java.net.URI;
+import java.time.ZonedDateTime;
+
+@FunctionalInterface
+public interface CloudEventAttributesVisitor {
```

Comment:
No, the visitor won't change. It is designed to be forward and backward compatible and a good implementation of CloudEvent must implement it in order to access to the other sdk features. So this really belongs to api and not to core

Reply:
My opinion is that visitor is just one mechanism which may or may not be the best for a given task. And yes I agree with Sergei that it does not belong here, but given that the real value of this PR is the fact that it accepts the changes proposed by #143 (streamlining CloudEvent interface), i say we can debate the placement of the visitor interface some other time.

Reply:
@slinkydeveloper 

>  It is designed to be forward and backward compatible and a good implementation of CloudEvent must implement it in order to access to the other sdk features

After looking at `-api`, I got a feeling that I can implement everything just by using `CloudEvents` interface. That makes me think that the visitors are optional, or even a detail of the implementation of `core`.

What I suggest here is to put them into `core`, and when/if we feel that they make sense exposing publicly as part of the API, we move them there as a follow up. WDYT? :)

Reply:
I think it's not a good idea, as i clearly underlined in my other proposals and yesterday meeting.

Reply:
If in the process of vendoring the sdk you think these interfaces are useless, feel free to remove them from the classpath.

Reply:
> If in the process of vendoring the sdk you think these interfaces are useless, feel free to remove them from the classpath using one of the various maven plugins designed for this purpose.

that's a dependency, one cannot alter dependencies or anything.

But okay, just wanted to state that these visitors are useless for those who only want to implement `CloudEvents` interface (e.g. by wrapping an existing type), so it really becomes a detail of the implementation of `core`.

Since there are no references to the visitors in `CloudEvents` interface, It does not block me, I can just choose not to provide any implementation for them 👍

Reply:
You can always vendor a dependency shipping these interfaces with your jar (eg spring-cloudevents vendors directly cloudevents-api shipping a jar with these interfaces inside its classpath). In that case, you can remove classpath elements using the maven-jar-plugin.

Reply:
@slinkydeveloper we don't do this in Java. This will also break classpaths of some projects.